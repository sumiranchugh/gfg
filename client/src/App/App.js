import React, { Component } from "react";
import Header from "../Header/Header";
import Search from "../Search/Search";
import Products from "../Product/Products";
import { Container, Col, Row, Alert } from "react-bootstrap";
import { fetchProductsAPI } from "../common/apis/productApi";

class App extends Component {
  constructor() {
    super();
    this.state = {
      products: [],
      error: { state: false, message: "" }
    };
  }

  searchCallback = newProducts => this.setState({ products: newProducts });
  errorCallback = err => this.setState({ error: err });
  render() {
    return (
      <div className="App">
        <Header />
        <Container fluid>
          <Row className="justify-content-center mt-3">
            <Col lg="6">
              <Search
                updateProducts={this.searchCallback}
                errorCallback={this.errorCallback}
              />
            </Col>
          </Row>
          <Row className="justify-content-center">
            {this.state.error.status ? (
              <Alert className="mt-5" variant="danger">
                <Alert.Heading>{this.state.error.message}</Alert.Heading>
                Looks like we've got some trouble!. Please try again later.
              </Alert>
            ) : (
              <Products products={this.state.products} />
            )}
          </Row>
        </Container>
      </div>
    );
  }
  componentDidMount() {
    fetchProductsAPI(
      data => this.setState({ products: data }),
      err => this.setState({ error: err })
    );
  }
}

export default App;
