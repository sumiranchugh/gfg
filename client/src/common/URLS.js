export const PRODUCTS = "api/products/";
export const PRODUCTS_FILTER = "api/products/filter?q=";
