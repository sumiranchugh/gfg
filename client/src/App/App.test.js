import React from "react";
import Adapter from "enzyme-adapter-react-16";
import renderer from "react-test-renderer";
import App from "./App";
import expect from "expect";
import { shallow, mount, render, configure } from "enzyme";
import { PRODUCTS_FILTER, PRODUCTS } from "../common/URLS";
configure({ adapter: new Adapter() });

// it("renders without crashing", () => {
//   const tree = renderer.create(<App/>).toJSON();
//   expect(tree).toMatchSnapshot();
// });

describe("App", () => {
  it("render app with no products", done => {
    const mockSuccessResponse = [];
    const mockJsonPromise = Promise.resolve(mockSuccessResponse);
    const mockFetchPromise = Promise.resolve({
      json: () => mockJsonPromise
    });
    global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);
    const wrapper = shallow(<App />);
    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(global.fetch).toHaveBeenCalledWith(PRODUCTS);

    expect(wrapper).toMatchSnapshot();
    process.nextTick(() => {
      expect(wrapper.state().products).toEqual(mockSuccessResponse);
    })
    global.fetch.mockClear();
    delete global.fetch;
    done();
  });
  it("render app with 1 products", done => {
    const mockSuccessResponse = [{'id': 1,'name':'product1', 'price':20,'brand': 'gucci'}];
    const mockJsonPromise = Promise.resolve(mockSuccessResponse);
    const mockFetchPromise = Promise.resolve({
      json: () => mockJsonPromise
    });
    global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);
    const wrapper = shallow(<App />);
    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(global.fetch).toHaveBeenCalledWith(PRODUCTS);
    process.nextTick(() => {
      expect(wrapper.state().products).toEqual(mockSuccessResponse);
      // expect(wrapper.find('Products').props().products).toEqual(mockSuccessResponse);
    })
    
    expect(wrapper).toMatchSnapshot();
    // wrapper.unmount();
    // console.log(wrapper.debug())
    global.fetch.mockClear();
    delete global.fetch;
    
    done();
  });
  it("render app with backend error", done => {
    global.fetch = jest.fn().mockRejectedValue('');
    const wrapper = shallow(<App />);
    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(global.fetch).toHaveBeenCalledWith(PRODUCTS);
    process.nextTick(() => {
      expect(wrapper.state().products).toEqual([]);
      expect(wrapper.state().error.state).toEqual(true);
    })
    expect(wrapper).toMatchSnapshot();
    global.fetch.mockClear();
    delete global.fetch;
    done();
  });
});
