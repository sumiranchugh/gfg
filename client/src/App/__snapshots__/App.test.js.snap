// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App render app with 1 products 1`] = `ShallowWrapper {}`;

exports[`App render app with backend error 1`] = `ShallowWrapper {}`;

exports[`App render app with no products 1`] = `ShallowWrapper {}`;
